Model/ProductProtection.php

```diff
@@ -7,94 +7,227 @@
+        string $leadToken = null,
+        string $listPrice = null,
+        string $orderOfferPlanId = null
+    ): void {
+        try {
+            $product = $this->productRepository->get(Extend::WARRANTY_PRODUCT_SKU);
+            $quote = $this->checkoutSession->getQuote();
+            $quoteId = $quote->getId();
 
-        $params = array();
-        $options = array();
-        $params['qty'] = $quantity;
-        $params['product'] = $product->getId();
-        $product->setData('extend_plan_price', ($price/100));
+            if ($price === 0) {
+                throw new Exception('Cannot add/update product protection with a price of 0');
+            }
 
+            // if quantity is 0, remove the item from the quote
+            if ($quantity === 0 && isset($cartItemId)) {
+                $quote->removeItem($cartItemId);
+                $this->quoteRepository->save($quote);
+                return;
+            }
+
+            if (isset($cartItemId)) {
+                $item = $this->checkoutSession->getQuote()->getItemById($cartItemId);
+            }
+
+            // if we are adding pp, or we didn't find an existing item, create a new one
+            if (!isset($item) || $item === false) {
+                // ensure that we have the required properties to create the protection plan
+                if (
+                    !isset($quantity) ||
+                    !isset($productId) ||
+                    !isset($planId) ||
+                    !isset($price) ||
+                    !isset($term) ||
+                    !isset($coverageType)
+                ) {
+                    throw new Exception(
+                        'Missing required parameters to add product protection to cart.'
+                    );
+                }
+                $item = $this->itemFactory->create();
+            }
+
+            // if quote has not been saved yet, save it so that we have an id
+            if (!isset($quoteId)) {
+                $this->quoteRepository->save($quote);
+                $quote = $this->checkoutSession->getQuote();
+                $quoteId = $quote->getId();
+            }
+
+            $item->setQuoteId($quoteId);
+            $item->setProduct($product);
+
+            if (isset($quantity)) {
+                $item->setQty($quantity);
+            }
+
+            if (isset($price)) {
+                $item
+                    ->setCustomPrice($price / 100)
+                    ->setOriginalCustomPrice($price / 100)
+                    ->getProduct()
+                    ->setIsSuperMode(true);
+            }
+
+            $options = $this->createOptions($product, $item, [
+                'Plan Type' => $coverageType,
+                'Term' => $term,
+                'List Price' => $listPrice,
+                'Order Offer Plan Id' => $orderOfferPlanId,
+                'Lead Token' => $leadToken,
+                'Associated Product' => $productId,
+                'Plan ID' => $planId,
```

Comment:
Is it possible that an _update_ is requested but some of these variable values are `null`?

Reply:
Yes, we grab the existing values off of the item inside of the createOptions func so if they are null we fall back to the existing option, and if it doesn't exist it just doesn't get added.